nmigen_stdio/spiflash.py

```diff
@@ -0,0 +1,425 @@
+                             .format(protocol))
+        self._protocol = protocol
+        self._data_width = data_width
+
+        if divisor < 1:
+            raise ValueError("Invalid divisor value; must be at least 1"
+                             .format(divisor))
+        self.divisor = Signal(bits_for(divisor), reset=divisor)
+        self._divisor_val = divisor + 1
+
+        supported_devices = ["lattice_ecp5"]
+        if device is not None and device not in supported_devices:
+            raise ValueError("Invalid FPGA device name {!r}; must be one of {}"
+                             .format(device, supported_devices))
+        self._device = device
+        if self._device is not None and pins is None:
+            raise ValueError("Pins parameter is missing for this FPGA device {}"
+                             .format(self._device))
+        self._pins = pins
+
+        if self._protocol == "extended":
+            self.spi_width = 1
+        elif self._protocol == "dual":
+            self.spi_width = 2
+        elif self._protocol == "quad":
+            self.spi_width = 4
+
+        self.cs = Signal(reset=0)       # Equivalent to ~CS_N
+        self.clk = Signal()
+
+        if protocol == "extended":
+            self.mosi = Signal()
+            self.miso = Signal()
+        elif protocol == "dual":
+            self.dq = Pin(2, "io")
+        elif protocol == "quad":
+            self.dq = Pin(4, "io")
+
+        self.rdy = Signal()             # Internal rdy signal, set at idle state
+        self.ack = Signal()             # External ack signal that begins a read
+        self.r_data = Signal(self._data_width)
+        self.r_rdy = Signal()           # 1 if r_data is valid, 0 otherwise
+
+        self.cmd_width = type(self).CMD_WIDTH
+        self._fcmd_width = self.cmd_width * self.spi_width
+        # A two-way register storing the current value on the DQ I/O pins
+        # (Note: DQ pins are both input/output only for Dual or Quad SPI protocols)
+        self.shreg = Signal(max(self._fcmd_width, self._data_width))
+        self.counter = Signal.like(self.divisor)
+
+
+    def _add_clk_primitive(self, platform, module):
+        """Add a submodule whose instantiation is required by certain devices
+        when choosing a user clock as SPI clock.
+
+        Note that the CLK pin is not connected outside of this function, 
+        as certain devices does not need to request the CLK pin.
+        """
+        # Lattice ECP5:
+        # "The ECP5 and ECP5-5G devices provide a solution for users 
+        # to choose any user clock as MCLK under this scenario 
+        # by instantiating USRMCLK macro in your Verilog or VHDL."
+        # (see Section 6.1.2 of FPGA-TN-02039-1.7, 
+        #  "ECP5 and ECP5-5G sysCONFIG Usage Guide Technical Note")
+        if self._device == "lattice_ecp5":
+            module.submodules += Instance("USRMCLK",
```

Comment:
I think it is best to leave this out of the SPI core, and move it into nmigen-boards (failing that, in the top-level SoC file as it is currently done with MiSoC designs):
1. It makes this core simpler.
2. It makes it device-independent (there is nothing else AFAICS in this code that would not work on other FPGAs).
3. Other SPI flash cores could benefit from this automatic management of ``USRMCLK`` and other similar FPGA vendor inventions.
4. It allows this core to be used with flash chips that are not the configuration memory of the FPGA, and use other pins.
5. In some scenarios, the user may want other signals passed to those primitives. Leaving them out of the SPI flash core allows for more flexibility.